File: reactfrontend/src/services/EmployeeService.js
import axios from 'axios';

const API_URL = '/api/employees';

const handleError = (error, message) => {
  console.error(message, error);
  if (error.response?.data?.message) {
    throw new Error(error.response.data.message);
  }
  throw error;
};

const EmployeeService = {
  getAllEmployees: async () => {
    try {
      const response = await axios.get(`${API_URL}/all`);
      return response.data;
    } catch (error) {
      console.error('Error fetching all employees:', error);
      throw error;
    }
  },
  
  getActiveEmployees: async () => {
    try {
      const response = await axios.get(API_URL);
      return response.data;
    } catch (error) {
      console.error('Error fetching active employees:', error);
      throw error;
    }
  },
  
  getEmployeeDetails: async (id) => {
    try {
      const response = await axios.get(`${API_URL}/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching employee details for ID ${id}:`, error);
      throw error;
    }
  },
  
  createEmployee: async (employee, managerId) => {
    try {
      const url = managerId ? `${API_URL}?managerId=${managerId}` : API_URL;
      // Backend expects dates in MM/dd/yyyy format
      const employeeData = {
        ...employee,
        hireDate: employee.hireDate // hireDate should already be in MM/dd/yyyy format from the form
      };
      const response = await axios.post(url, employeeData);
      return response.data;
    } catch (error) {
      return handleError(error, 'Error creating employee');
    }
  },
  
  deactivateEmployee: async (id) => {
    try {
      const response = await axios.put(`${API_URL}/${id}/deactivate`);
      return response.data;
    } catch (error) {
      return handleError(error, `Error deactivating employee with ID ${id}`);
    }
  },
  
  activateEmployee: async (id) => {
    try {
      const response = await axios.put(`${API_URL}/${id}/reactivate`);
      return response.data;
    } catch (error) {
      return handleError(error, `Error activating employee with ID ${id}`);
    }
  },
  
  getEmployeesByHireDate: async (startDate, endDate) => {
    try {
      // Validate dates
      const today = '2025-04-13';
      const minDate = '1925-04-13';

      if (startDate > today || endDate > today) {
        throw new Error('Cannot search with future dates');
      }

      if (startDate < minDate || endDate < minDate) {
        throw new Error('Dates cannot be more than 100 years in the past');
      }

      if (startDate > endDate) {
        throw new Error('Start date must be before or equal to end date');
      }

      console.log('Sending date range request:', { startDate, endDate });
      const response = await axios.get(`${API_URL}/hired`, {
        params: {
          startDate: startDate,
          endDate: endDate
        }
      });
      console.log('Got response:', response.data);
      return response.data;
    } catch (error) {
      console.error('Error fetching employees by hire date:', error);
      throw error;
    }
  },



  deleteEmployee: async (id) => {
    console.log(`Attempting to delete employee with ID: ${id}`);
    try {
      const response = await axios.delete(`${API_URL}/${id}`);
      console.log(`Successfully deleted employee with ID: ${id}`, response);
    } catch (error) {
      console.error(`Failed to delete employee with ID: ${id}`, error.response || error);
      return handleError(error, `Error deleting employee with ID ${id}`);
    }
  }
};

export default EmployeeService;

File: src/main/java/com/flywire/exercise/service/EmployeeService.java
package com.flywire.exercise.service;

import com.flywire.exercise.exception.EmployeeNotFoundException;
import com.flywire.exercise.exception.InvalidEmployeeDataException;
import com.flywire.exercise.model.Employee;
import com.flywire.exercise.util.FileUtil;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.stream.Collectors;
import java.util.TimeZone;

@Service
public class EmployeeService {

    private final FileUtil fileUtil;

    public EmployeeService(FileUtil fileUtil) {
        this.fileUtil = fileUtil;
    }

    public List<Employee> getAllActiveEmployees() throws IOException {
        List<Employee> employees = fileUtil.readEmployees();
        return employees.stream()
                .filter(Employee::isActive)
                .sorted(Comparator.comparing(e -> e.getName().split(" ")[1])) // Sort by last name
                .collect(Collectors.toList());
    }

    public List<Employee> getAllEmployees() throws IOException {
        List<Employee> employees = fileUtil.readEmployees();
        return employees.stream()
                .sorted(Comparator.comparing(e -> e.getName().split(" ")[1])) // Sort by last name
                .collect(Collectors.toList());
    }

    public Map<String, Object> getEmployeeWithDirectHires(Long id) throws IOException {
        List<Employee> employees = fileUtil.readEmployees();
        
        // Find the employee
        Employee employee = employees.stream()
                .filter(e -> e.getId().equals(id))
                .findFirst()
                .orElseThrow(() -> new EmployeeNotFoundException("Employee not found with id: " + id));

        // Get direct reports that still exist in the system
        List<Employee> directHires = new ArrayList<>();
        if (employee.getDirectReports() != null && !employee.getDirectReports().isEmpty()) {
            directHires = employees.stream()
                    .filter(e -> employee.getDirectReports().contains(e.getId()))
                    .collect(Collectors.toList());

            // Update employee's direct reports to only include existing employees
            employee.setDirectReports(directHires.stream()
                    .map(Employee::getId)
                    .collect(Collectors.toList()));
            fileUtil.writeEmployees(employees);
        }

        Map<String, Object> response = new HashMap<>();
        response.put("employee", employee);
        response.put("directHires", directHires);
        return response;
    }

    public List<Employee> getEmployeesByHireDateRange(Date startDate, Date endDate) throws IOException {
        SimpleDateFormat sdf = new SimpleDateFormat("MM/dd/yyyy");
        sdf.setTimeZone(TimeZone.getTimeZone("UTC"));

        // Validate dates
        if (startDate == null || endDate == null) {
            throw new InvalidEmployeeDataException("Both start date and end date are required");
        }

        try {
            // Normalize input dates
            startDate = sdf.parse(sdf.format(startDate));
            endDate = sdf.parse(sdf.format(endDate));

            if (startDate.after(endDate)) {
                throw new InvalidEmployeeDataException("Start date must be before end date");
            }

            System.out.println("Searching for employees between " + sdf.format(startDate) + " and " + sdf.format(endDate));

            // Get all employees and filter by hire date
            List<Employee> employees = fileUtil.readEmployees();
            System.out.println("Total employees before filtering: " + employees.size());

            List<Employee> filteredEmployees = new ArrayList<>();
            
            for (Employee e : employees) {
                if (e.getHireDate() == null) {
                    System.out.println("Skipping employee " + e.getName() + ": no hire date");
                    continue;
                }

                // Normalize hire date
                Date hireDate = sdf.parse(sdf.format(e.getHireDate()));
                System.out.println("Checking employee " + e.getName() + " with hire date: " + sdf.format(hireDate));

                if ((hireDate.equals(startDate) || hireDate.after(startDate)) && 
                    (hireDate.equals(endDate) || hireDate.before(endDate))) {
                    System.out.println("Including employee: " + e.getName());
                    filteredEmployees.add(e);
                } else {
                    System.out.println("Excluding employee: " + e.getName() + 
                                     " (hire date outside range: " + sdf.format(hireDate) + ")");
                }
            }

            // Sort by hire date descending
            filteredEmployees.sort((e1, e2) -> e2.getHireDate().compareTo(e1.getHireDate()));

            System.out.println("Found " + filteredEmployees.size() + " employees in date range");
            return filteredEmployees;

        } catch (ParseException ex) {
            System.err.println("Error parsing dates: " + ex.getMessage());
            throw new InvalidEmployeeDataException("Error processing dates");
        }
    }

    public Employee createEmployee(Employee employee, Long managerId) throws IOException {
        List<Employee> employees = fileUtil.readEmployees();

        // Validate required fields
        if (employee.getName() == null || employee.getName().trim().isEmpty()) {
            throw new InvalidEmployeeDataException("Employee name is required");
        }
        if (employee.getPosition() == null || employee.getPosition().trim().isEmpty()) {
            throw new InvalidEmployeeDataException("Employee position is required");
        }
        if (employee.getHireDate() == null) {
            throw new InvalidEmployeeDataException("Hire date is required");
        }
        // Normalize dates for comparison
        Calendar cal = Calendar.getInstance();
        cal.set(Calendar.HOUR_OF_DAY, 0);
        cal.set(Calendar.MINUTE, 0);
        cal.set(Calendar.SECOND, 0);
        cal.set(Calendar.MILLISECOND, 0);
        Date today = cal.getTime();

        cal.setTime(employee.getHireDate());
        cal.set(Calendar.HOUR_OF_DAY, 0);
        cal.set(Calendar.MINUTE, 0);
        cal.set(Calendar.SECOND, 0);
        cal.set(Calendar.MILLISECOND, 0);
        employee.setHireDate(cal.getTime());

        if (employee.getHireDate().after(today)) {
            throw new InvalidEmployeeDataException("Hire date cannot be in the future");
        }

        // Name validation
        String name = employee.getName().trim();
        if (name.isEmpty()) {
            throw new InvalidEmployeeDataException("Name is required");
        }

        // Date format validation
        try {
            SimpleDateFormat dateFormat = new SimpleDateFormat("MM/dd/yyyy");
            dateFormat.setLenient(false);
            String formattedDate = dateFormat.format(employee.getHireDate());
            Date parsedDate = dateFormat.parse(formattedDate);
            employee.setHireDate(parsedDate);
        } catch (ParseException e) {
            throw new InvalidEmployeeDataException("Invalid hire date format. Expected format: MM/DD/YYYY");
        }

        // Validate manager
        if (managerId != null) {
            Employee manager = employees.stream()
                    .filter(e -> e.getId().equals(managerId))
                    .findFirst()
                    .orElseThrow(() -> new EmployeeNotFoundException("Manager not found with id: " + managerId));

            if (!manager.isActive()) {
                throw new InvalidEmployeeDataException("Cannot assign an inactive employee as manager");
            }

            // Add employee to manager's direct reports
            List<Long> reports = new ArrayList<>(manager.getDirectReports());
            reports.add(employee.getId());
            manager.setDirectReports(reports);
        }

        // Set default values
        employee.setActive(true);
        if (employee.getDirectReports() == null) {
            employee.setDirectReports(new ArrayList<>());  // Only set empty list if null
        }
        employee.setOriginal(false);  // Mark as a new employee

        // Validate direct reports if provided
        if (!employee.getDirectReports().isEmpty()) {
            for (Long reportId : employee.getDirectReports()) {
                if (!employees.stream().anyMatch(e -> e.getId().equals(reportId))) {
                    throw new InvalidEmployeeDataException("Direct report not found with id: " + reportId);
                }
            }
        }

        // Validate that the ID doesn't already exist
        if (employee.getId() == null) {
            throw new InvalidEmployeeDataException("Employee ID is required");
        }
        
        boolean idExists = employees.stream()
                .anyMatch(e -> e.getId().equals(employee.getId()));
        if (idExists) {
            throw new InvalidEmployeeDataException("Employee ID " + employee.getId() + " already exists");
        }

        employees.add(employee);
        fileUtil.writeEmployees(employees);
        return employee;
    }

    public Employee deactivateEmployee(Long id) throws IOException {
        List<Employee> employees = fileUtil.readEmployees();
        Employee employee = employees.stream()
                .filter(e -> e.getId().equals(id))
                .findFirst()
                .orElseThrow(() -> new EmployeeNotFoundException("Employee not found with id: " + id));

        if (!employee.isActive()) {
            throw new InvalidEmployeeDataException("Employee is already inactive");
        }

        // Remove this employee from their manager's direct reports
        employees.stream()
                .filter(e -> e.getDirectReports().contains(id))
                .findFirst()
                .ifPresent(manager -> {
                    List<Long> reports = new ArrayList<>(manager.getDirectReports());
                    reports.remove(id);
                    manager.setDirectReports(reports);
                });

        // Deactivate the employee
        employee.setActive(false);
        employee.setDirectReports(new ArrayList<>()); // Clear direct reports when deactivating

        fileUtil.writeEmployees(employees);
        return employee;
    }

    public Employee reactivateEmployee(Long id) throws IOException {
        List<Employee> employees = fileUtil.readEmployees();
        Employee employee = employees.stream()
                .filter(e -> e.getId().equals(id))
                .findFirst()
                .orElseThrow(() -> new EmployeeNotFoundException("Employee not found with id: " + id));

        if (employee.isActive()) {
            throw new InvalidEmployeeDataException("Employee is already active");
        }

        employee.setActive(true);
        fileUtil.writeEmployees(employees);
        return employee;
    }

    public void deleteEmployee(Long id) throws IOException {
        List<Employee> employees = fileUtil.readEmployees();
        Employee employee = employees.stream()
                .filter(e -> e.getId().equals(id))
                .findFirst()
                .orElseThrow(() -> new EmployeeNotFoundException("Employee not found with id: " + id));

        // Check if employee is original
        if (employee.isOriginal()) {
            throw new InvalidEmployeeDataException("Cannot delete original employees");
        }

        // Remove this employee from their manager's direct reports
        employees.stream()
                .filter(e -> e.getDirectReports().contains(id))
                .findFirst()
                .ifPresent(manager -> {
                    List<Long> reports = new ArrayList<>(manager.getDirectReports());
                    reports.remove(id);
                    manager.setDirectReports(reports);
                });

        // Remove the employee
        employees.removeIf(e -> e.getId().equals(id));
        fileUtil.writeEmployees(employees);
    }
}
